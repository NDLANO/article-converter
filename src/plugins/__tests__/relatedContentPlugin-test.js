/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import bunyan from 'bunyan';
import log from '../../utils/logger';
import createRelatedContentPlugin from '../relatedContentPlugin';

const articleResource = [
  {
    resourceTypes: [
      {
        id: 'urn:resourcetype:tasksAndActivities',
        name: 'Oppgaver og aktiviteter',
      },
    ],
    path: '/subject:12/topic:1:183846/topic:1:183935/resource:1:110269',
  },
];

test('fetchResource when no article ids is provided', async () => {
  const relatedContentPlugin = createRelatedContentPlugin();
  const resource = await relatedContentPlugin.fetchResource({ data: {} });
  expect(resource).toEqual({ data: {} });
});

test('fetchResource for two related articles', async () => {
  const relatedContentPlugin = createRelatedContentPlugin();

  ['1', '2'].forEach(id => {
    nock('http://ndla-api')
      .get(`/article-api/v2/articles/${id}?language=nb&fallback=true`)
      .reply(200, {
        title: { title: `title${id}` },
        introduction: { introduction: `introduction${id}` },
      });
    nock('http://ndla-api')
      .get(
        `/taxonomy/v1/queries/resources?contentURI=urn:article:${id}&language=nb`
      )
      .reply(200, articleResource);
  });

  const resource1 = await relatedContentPlugin.fetchResource(
    {
      data: { articleId: '1' },
    },
    'token',
    'nb'
  );
  expect(resource1).toMatchSnapshot();

  const resource2 = await relatedContentPlugin.fetchResource(
    {
      data: { articleId: '2' },
    },
    'token',
    'nb'
  );
  expect(resource2).toMatchSnapshot();
});

test('fetchResource for an external article', async () => {
  log.level(bunyan.FATAL + 1); // temporarily disable logging

  const relatedContentPlugin = createRelatedContentPlugin();

  const externalResource = await relatedContentPlugin.fetchResource(
    {
      data: {
        title: 'Helsedirektoratet om reklame for alkohol',
        url:
          'https://helsedirektoratet.no/folkehelse/alkohol/forbud-mot-alkoholreklame',
      },
    },
    'token',
    'nb'
  );
  expect(externalResource).toMatchSnapshot();

  log.level(bunyan.INFO);
});

test('fetchResource for two related articles, where one could not be fetched from article-api', async () => {
  log.level(bunyan.FATAL + 1); // temporarily disable logging

  const relatedContentPlugin = createRelatedContentPlugin();
  nock('http://ndla-api')
    .get(`/article-api/v2/articles/1?language=nb&fallback=true`)
    .reply(200, {
      title: { title: `title1` },
      introduction: { introduction: `introduction1` },
    });

  nock('http://ndla-api')
    .get(`/taxonomy/v1/queries/resources?contentURI=urn:article:1&language=nb`)
    .reply(200, articleResource);

  nock('http://ndla-api')
    .get(`/article-api/v2/articles/2?language=nb&fallback=true`)
    .reply(501, {});

  const resource1 = await relatedContentPlugin.fetchResource(
    {
      data: { articleId: '1' },
    },
    'token',
    'nb'
  );
  expect(resource1).toMatchSnapshot();

  const resource2 = await relatedContentPlugin.fetchResource(
    {
      data: { articleId: '2' },
    },
    'token',
    'nb'
  );
  expect(resource2).toMatchSnapshot();

  log.level(bunyan.INFO);
});

test('fetchResource for two related articles, where one could not be fetched from taxonomy-api', async () => {
  log.level(bunyan.FATAL + 1); // temporarily disable logging

  const relatedContentPlugin = createRelatedContentPlugin();
  nock('http://ndla-api')
    .get(`/article-api/v2/articles/1?language=nb&fallback=true`)
    .reply(200, {
      title: { title: `title1` },
      introduction: { introduction: `introduction1` },
    });
  nock('http://ndla-api')
    .get(`/taxonomy/v1/queries/resources?contentURI=urn:article:1&language=nb`)
    .reply(200, articleResource);

  nock('http://ndla-api')
    .get(`/article-api/v2/articles/2?language=nb&fallback=true`)
    .reply(200, {
      title: { title: `title2` },
      introduction: { introduction: `introduction2` },
    });
  nock('http://ndla-api')
    .get(`/taxonomy/v1/queries/resources?contentURI=urn:article:2&language=nb`)
    .reply(500, {});

  const resource1 = await relatedContentPlugin.fetchResource(
    {
      data: { articleId: '1' },
    },
    'token',
    'nb'
  );
  expect(resource1).toMatchSnapshot();

  const resource2 = await relatedContentPlugin.fetchResource(
    {
      data: { articleId: '2' },
    },
    'token',
    'nb'
  );
  expect(resource2).toMatchSnapshot();

  log.level(bunyan.INFO);
});

test('embedToHtml should return empty string if no related articles is provided', async () => {
  const relatedContentPlugin = createRelatedContentPlugin();

  expect(relatedContentPlugin.embedToHTML({ embed: {} })).toBe('');
  expect(relatedContentPlugin.embedToHTML({ embed: { article: {} } })).toBe('');
});

test('embedToHtml should return fallback url if no resource was found', async () => {
  const relatedContentPlugin = createRelatedContentPlugin();

  const embed1 = {
    data: { articleId: '1231', resource: 'related-content' },
    article: {
      id: 1231,
      title: { title: 't1' },
      metaDescription: { metaDescription: 'm1' },
      resource: {},
    },
  };
  expect(relatedContentPlugin.embedToHTML(embed1)).toMatchSnapshot();

  const embed2 = {
    data: { articleId: '1145', resource: 'related-content' },
    article: {
      id: 1145,
      title: { title: 't2' },
      metaDescription: { metaDescription: 'm2' },
      resource: {
        path: '/subject:4/topic:1:172816/topic:1:178048/resource:1:74420',
        resourceTypes: [
          { id: 'urn:resourcetype:academicArticle', name: 'Fagartikkel' },
          { id: 'urn:resourcetype:subjectMaterial', name: 'Fagstoff' },
        ],
      },
    },
  };
  expect(relatedContentPlugin.embedToHTML(embed2)).toMatchSnapshot();
});
