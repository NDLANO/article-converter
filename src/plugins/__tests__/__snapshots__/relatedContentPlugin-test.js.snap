// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embedToHtml should return fallback url if no resource was found 1`] = `"<section class=\\"c-related-articles\\"><h1 class=\\"c-related-articles__component-title\\">Relaterte artikler</h1><div class=\\"c-related-articles__articles\\"><article class=\\"c-related-articles__item c-related-articles__item--subject-material\\"><h1 class=\\"c-related-articles__title\\"><div class=\\"c-content-type-badge c-content-type-badge--subject-material c-content-type-badge--small c-content-type-badge--background\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"1em\\" width=\\"1em\\" class=\\"c-icon\\" viewBox=\\"0 0 24 24\\" data-license=\\"Apache License 2.0\\" data-source=\\"Material Design\\" style=\\"vertical-align:middle\\"><g><title>Fagstoff</title><g id=\\"Page-1\\" stroke=\\"none\\" fill=\\"none\\" stroke-width=\\"1\\" fill-rule=\\"evenodd\\"><g id=\\"fagstoff\\"><g><polygon id=\\"Shape\\" points=\\"0 0 24 0 24 24 0 24\\"></polygon><path d=\\"M0,10 L0,8 L11,8 L11,10 L0,10 Z M24,0 L24,2 L0,2 L0,0 L24,0 Z M24,4 L24,6 L0,6 L0,4 L24,4 Z M13,8 L24,15.5 L13,23 L13,8 Z\\" id=\\"shape\\" fill=\\"currentColor\\"></path></g></g></g></g></svg></div><a href=\\"/article/1231\\" class=\\"c-related-articles__link\\">t1</a></h1><p class=\\"c-related-articles__description\\">m1</p></article><article class=\\"c-related-articles__item c-related-articles__item--subject-material\\"><h1 class=\\"c-related-articles__title\\"><div class=\\"c-content-type-badge c-content-type-badge--subject-material c-content-type-badge--small c-content-type-badge--background\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"1em\\" width=\\"1em\\" class=\\"c-icon\\" viewBox=\\"0 0 24 24\\" data-license=\\"Apache License 2.0\\" data-source=\\"Material Design\\" style=\\"vertical-align:middle\\"><g><title>Fagstoff</title><g id=\\"Page-1\\" stroke=\\"none\\" fill=\\"none\\" stroke-width=\\"1\\" fill-rule=\\"evenodd\\"><g id=\\"fagstoff\\"><g><polygon id=\\"Shape\\" points=\\"0 0 24 0 24 24 0 24\\"></polygon><path d=\\"M0,10 L0,8 L11,8 L11,10 L0,10 Z M24,0 L24,2 L0,2 L0,0 L24,0 Z M24,4 L24,6 L0,6 L0,4 L24,4 Z M13,8 L24,15.5 L13,23 L13,8 Z\\" id=\\"shape\\" fill=\\"currentColor\\"></path></g></g></g></g></svg></div><a href=\\"/subjects/subject:4/topic:1:172816/topic:1:178048/resource:1:74420\\" class=\\"c-related-articles__link\\">t2</a></h1><p class=\\"c-related-articles__description\\">m2</p></article></div></section>"`;

exports[`fetchResource for two related articles 1`] = `
Object {
  "articles": Array [
    Object {
      "introduction": Object {
        "introduction": "introduction1",
      },
      "resource": Object {
        "path": "/subject:12/topic:1:183846/topic:1:183935/resource:1:110269",
        "resourceTypes": Array [
          Object {
            "id": "urn:resourcetype:tasksAndActivities",
            "name": "Oppgaver og aktiviteter",
          },
        ],
      },
      "title": Object {
        "title": "title1",
      },
    },
    Object {
      "introduction": Object {
        "introduction": "introduction2",
      },
      "resource": Object {
        "path": "/subject:12/topic:1:183846/topic:1:183935/resource:1:110269",
        "resourceTypes": Array [
          Object {
            "id": "urn:resourcetype:tasksAndActivities",
            "name": "Oppgaver og aktiviteter",
          },
        ],
      },
      "title": Object {
        "title": "title2",
      },
    },
  ],
  "data": Object {
    "articleIds": "1,2",
  },
}
`;

exports[`fetchResource for two related articles, where one could not be fetched from article-api 1`] = `
Object {
  "articles": Array [
    Object {
      "introduction": Object {
        "introduction": "introduction1",
      },
      "resource": Object {
        "path": "/subject:12/topic:1:183846/topic:1:183935/resource:1:110269",
        "resourceTypes": Array [
          Object {
            "id": "urn:resourcetype:tasksAndActivities",
            "name": "Oppgaver og aktiviteter",
          },
        ],
      },
      "title": Object {
        "title": "title1",
      },
    },
  ],
  "data": Object {
    "articleIds": "1,2",
  },
}
`;

exports[`fetchResource for two related articles, where one could not be fetched from taxonomy-api 1`] = `
Object {
  "articles": Array [
    Object {
      "introduction": Object {
        "introduction": "introduction1",
      },
      "resource": Object {
        "path": "/subject:12/topic:1:183846/topic:1:183935/resource:1:110269",
        "resourceTypes": Array [
          Object {
            "id": "urn:resourcetype:tasksAndActivities",
            "name": "Oppgaver og aktiviteter",
          },
        ],
      },
      "title": Object {
        "title": "title1",
      },
    },
    Object {
      "introduction": Object {
        "introduction": "introduction2",
      },
      "resource": undefined,
      "title": Object {
        "title": "title2",
      },
    },
  ],
  "data": Object {
    "articleIds": "1,2",
  },
}
`;
